File: lab10/src/by/bntu/fitr/minigame/domain/west/ForestElf.java
package by.bntu.fitr.minigame.domain.west;

import by.bntu.fitr.minigame.domain.Creature;
import by.bntu.fitr.minigame.domain.Role;
import by.bntu.fitr.minigame.domain.Weapon;

public class ForestElf extends Creature {
    private Weapon weapon = Weapon.BOW;
    private Role role = Role.HUNTER;

    public ForestElf(int health, int strength, int agility, int intelligence) {
        super(health, strength, agility, intelligence);
    }

    @Override
    protected int getWeaponDamage() {
        return weapon.getDamage();
    }

    @Override
    public int buff() {
        return role.getPercent() / 100 * getAgility();
    }

    @Override
    public String toString() {
        return "ForestElf{" + super.toString() +
                "weapon=" + weapon +
                ", role=" + role +
                '}';
    }
}

File: lab10kt/src/main/kotlin/by/bntu/fitr/povt/korotkevich/model/west/ForestElf.kt
package by.bntu.fitr.povt.korotkevich.model.west

import by.bntu.fitr.povt.korotkevich.model.Creature
import by.bntu.fitr.povt.korotkevich.model.Role
import by.bntu.fitr.povt.korotkevich.model.Weapon
import by.bntu.fitr.povt.korotkevich.model.buff

class ForestElf(override val health: Int = 100, override val strength: Int = 20,
                override val agility: Int = 50, override val intelligence: Int = 20,
                val role: Role = Role.RANGER, val weapon: Weapon = Weapon.BOW) : Creature() {

    override val id: String = "ForestElf"

    override fun getSelfDamage() = weapon.damage + buff(role)

    override fun toString() = StringBuilder("ForestElf(").apply {
        append(super.toString())
        append("role: $role, weapon: $weapon)")
    }.toString()
}